File: plugins/plugin-cron/src/actions/create-cron.ts
/**
 * @module create-cron
 * @description Action to create a new cron job
 *
 * Supports natural language and structured input:
 * - "Create a cron job that runs every hour to check the weather"
 * - "Schedule a daily reminder at 9am"
 * - Structured JSON payload for programmatic use
 */

import type { Action, IAgentRuntime, Memory, State, HandlerCallback } from '@elizaos/core';
import { CronService } from '../services/cron-service.js';
import { CRON_SERVICE_TYPE, CronActions } from '../constants.js';
import type { CronJobCreate, CronSchedule, CronPayload } from '../types.js';
import {
  parseScheduleDescription,
  formatSchedule,
  validateSchedule,
} from '../scheduler/schedule-utils.js';
import { DEFAULT_CRON_CONFIG } from '../types.js';

/**
 * Parses a natural language cron request into components
 */
function parseNaturalLanguageRequest(text: string): {
  name?: string;
  schedule?: CronSchedule;
  prompt?: string;
  description?: string;
} {
  const result: {
    name?: string;
    schedule?: CronSchedule;
    prompt?: string;
    description?: string;
  } = {};

  const normalized = text.toLowerCase();

  // Try to extract schedule from various patterns
  // Pattern: "every X" (interval)
  const everyMatch = /every\s+(\d+\s*(?:second|minute|hour|day|week)s?)/i.exec(text);
  if (everyMatch) {
    result.schedule = parseScheduleDescription(`every ${everyMatch[1]}`);
  }

  // Pattern: "at X" (one-time or cron)
  const atMatch = /at\s+(\d{1,2}(?::\d{2})?\s*(?:am|pm)?)/i.exec(text);
  if (atMatch && !result.schedule) {
    // Convert time to cron expression
    const timeStr = atMatch[1].toLowerCase();
    let hours: number;
    let minutes = 0;

    const timeParts = /(\d{1,2})(?::(\d{2}))?\s*(am|pm)?/i.exec(timeStr);
    if (timeParts) {
      hours = parseInt(timeParts[1], 10);
      if (timeParts[2]) {
        minutes = parseInt(timeParts[2], 10);
      }
      if (timeParts[3]) {
        if (timeParts[3].toLowerCase() === 'pm' && hours !== 12) {
          hours += 12;
        } else if (timeParts[3].toLowerCase() === 'am' && hours === 12) {
          hours = 0;
        }
      }

      // Check for day patterns
      if (normalized.includes('daily') || normalized.includes('every day')) {
        result.schedule = {
          kind: 'cron',
          expr: `${minutes} ${hours} * * *`,
        };
      } else if (normalized.includes('weekday') || normalized.includes('monday to friday')) {
        result.schedule = {
          kind: 'cron',
          expr: `${minutes} ${hours} * * 1-5`,
        };
      } else if (normalized.includes('weekend')) {
        result.schedule = {
          kind: 'cron',
          expr: `${minutes} ${hours} * * 0,6`,
        };
      } else {
        // Default to daily
        result.schedule = {
          kind: 'cron',
          expr: `${minutes} ${hours} * * *`,
        };
      }
    }
  }

  // Pattern: "in X" (one-time delay)
  const inMatch = /in\s+(\d+\s*(?:second|minute|hour|day)s?)/i.exec(text);
  if (inMatch && !result.schedule) {
    result.schedule = parseScheduleDescription(`in ${inMatch[1]}`);
  }

  // Try to extract the action/prompt from "to X" or "that X"
  const toMatch = /(?:to|that)\s+(.+?)(?:\s+every|\s+at|\s+in\s+\d|$)/i.exec(text);
  if (toMatch) {
    result.prompt = toMatch[1].trim();
    // Use the prompt as the name if no explicit name
    result.name = toMatch[1].slice(0, 50).trim();
  }

  // Try to extract name from "called X" or "named X"
  const nameMatch = /(?:called|named)\s+["']?([^"']+)["']?/i.exec(text);
  if (nameMatch) {
    result.name = nameMatch[1].trim();
  }

  // If we still don't have a name, generate one
  if (!result.name && result.schedule) {
    const scheduleDesc = formatSchedule(result.schedule);
    result.name = `Cron job (${scheduleDesc})`;
  }

  return result;
}

/**
 * Formats a job for display
 */
function formatJobResponse(job: {
  id: string;
  name: string;
  schedule: CronSchedule;
  enabled: boolean;
  state: { nextRunAtMs?: number };
}): string {
  const scheduleStr = formatSchedule(job.schedule);
  const nextRun = job.state.nextRunAtMs
    ? new Date(job.state.nextRunAtMs).toLocaleString()
    : 'not scheduled';

  return (
    `Created cron job "${job.name}"\n` +
    `- ID: ${job.id}\n` +
    `- Schedule: ${scheduleStr}\n` +
    `- Status: ${job.enabled ? 'enabled' : 'disabled'}\n` +
    `- Next run: ${nextRun}`
  );
}

export const createCronAction: Action = {
  name: CronActions.CREATE_CRON,
  similes: [
    'SCHEDULE_CRON',
    'ADD_CRON',
    'NEW_CRON',
    'CREATE_SCHEDULED_JOB',
    'SET_UP_CRON',
    'SCHEDULE_JOB',
    'CREATE_RECURRING_JOB',
  ],
  description:
    'Creates a new cron job that runs on a schedule. Supports interval-based schedules (every X minutes), cron expressions, and one-time schedules.',

  validate: async (_runtime: IAgentRuntime, message: Memory): Promise<boolean> => {
    const text = message.content?.text?.toLowerCase() ?? '';

    // Check for scheduling-related keywords
    const hasScheduleKeyword =
      text.includes('cron') ||
      text.includes('schedule') ||
      text.includes('every ') ||
      text.includes('recurring') ||
      text.includes('repeat') ||
      text.includes('daily') ||
      text.includes('hourly') ||
      text.includes('weekly');

    // Check for creation intent
    const hasCreateIntent =
      text.includes('create') ||
      text.includes('add') ||
      text.includes('set up') ||
      text.includes('schedule') ||
      text.includes('make');

    return hasScheduleKeyword && hasCreateIntent;
  },

  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    _state?: State,
    options?: Record<string, unknown>,
    callback?: HandlerCallback
  ) => {
    // Get the cron service
    const cronService = runtime.getService<CronService>(CRON_SERVICE_TYPE);
    if (!cronService) {
      await callback?.({
        text: 'Cron service is not available. Please ensure the plugin is loaded.',
      });
      return { success: false, error: 'Cron service not available' };
    }

    const text = message.content?.text ?? '';

    // Check if this is a structured request (JSON in options)
    if (options?.jobInput && typeof options.jobInput === 'object') {
      const input = options.jobInput as CronJobCreate;

      // Validate the input
      const scheduleError = validateSchedule(input.schedule, DEFAULT_CRON_CONFIG);
      if (scheduleError) {
        await callback?.({
          text: `Invalid schedule: ${scheduleError}`,
        });
        return { success: false, error: scheduleError };
      }

      const job = await cronService.createJob(input);
      await callback?.({
        text: formatJobResponse(job),
      });
      return { success: true, data: { jobId: job.id, job } };
    }

    // Parse natural language request
    const parsed = parseNaturalLanguageRequest(text);

    if (!parsed.schedule) {
      await callback?.({
        text:
          "I couldn't understand the schedule. Please specify when the job should run, for example:\n" +
          '- "every 5 minutes"\n' +
          '- "every hour"\n' +
          '- "daily at 9am"\n' +
          '- "every weekday at 8:30am"',
      });
      return { success: false, error: 'Could not parse schedule' };
    }

    // Build the job input
    const jobInput: CronJobCreate = {
      name: parsed.name || 'Unnamed cron job',
      description: parsed.description,
      enabled: true,
      schedule: parsed.schedule,
      payload: {
        kind: 'prompt',
        text: parsed.prompt || 'Run scheduled task',
      } as CronPayload,
    };

    // If the user specified a one-time schedule, mark it for deletion after run
    if (parsed.schedule.kind === 'at') {
      jobInput.deleteAfterRun = true;
    }

    const job = await cronService.createJob(jobInput);

    await callback?.({
      text: formatJobResponse(job),
    });

    return {
      success: true,
      data: {
        jobId: job.id,
        job,
      },
    };
  },

  examples: [
    [
      {
        name: '{{user1}}',
        content: { text: 'Create a cron job to check the news every hour' },
      },
      {
        name: '{{agentName}}',
        content: {
          text: 'Created cron job "check the news"\n- ID: abc-123\n- Schedule: every 1 hour\n- Status: enabled\n- Next run: in 1 hour',
        },
      },
    ],
    [
      {
        name: '{{user1}}',
        content: { text: 'Schedule a daily reminder at 9am to review my goals' },
      },
      {
        name: '{{agentName}}',
        content: {
          text: 'Created cron job "review my goals"\n- ID: def-456\n- Schedule: cron: 0 9 * * *\n- Status: enabled\n- Next run: tomorrow at 9:00 AM',
        },
      },
    ],
    [
      {
        name: '{{user1}}',
        content: { text: 'Set up a recurring job every 5 minutes to check server status' },
      },
      {
        name: '{{agentName}}',
        content: {
          text: 'Created cron job "check server status"\n- ID: ghi-789\n- Schedule: every 5 minutes\n- Status: enabled\n- Next run: in 5 minutes',
        },
      },
    ],
  ],
};

File: plugins/plugin-cron/rust/src/actions/create_cron.rs
use async_trait::async_trait;
use serde_json::Value;

use crate::schedule::{format_schedule, parse_natural_language_schedule, parse_schedule};
use crate::types::{PayloadType, ScheduleType};
use crate::{Action, ActionExample, ActionResult, CronService};

pub struct CreateCronAction;

#[async_trait]
impl Action for CreateCronAction {
    fn name(&self) -> &str {
        "CREATE_CRON"
    }

    fn similes(&self) -> Vec<&str> {
        vec![
            "SCHEDULE_CRON",
            "ADD_CRON",
            "NEW_CRON",
            "CREATE_SCHEDULED_JOB",
            "SET_UP_CRON",
            "SCHEDULE_JOB",
        ]
    }

    fn description(&self) -> &str {
        "Creates a new cron job that runs on a schedule. Supports interval-based, cron expressions, and one-time schedules."
    }

    async fn validate(&self, message: &Value, _state: &Value) -> bool {
        let text = message
            .get("content")
            .and_then(|c| c.get("text"))
            .and_then(|t| t.as_str())
            .unwrap_or("")
            .to_lowercase();

        let has_schedule = text.contains("cron")
            || text.contains("schedule")
            || text.contains("every ")
            || text.contains("recurring")
            || text.contains("daily")
            || text.contains("hourly");

        let has_create = text.contains("create")
            || text.contains("add")
            || text.contains("set up")
            || text.contains("schedule")
            || text.contains("make");

        has_schedule && has_create
    }

    async fn handler(
        &self,
        message: &Value,
        _state: &Value,
        service: Option<&mut CronService>,
    ) -> ActionResult {
        let service = match service {
            Some(s) => s,
            None => {
                return ActionResult {
                    success: false,
                    text: "Cron service is not available.".to_string(),
                    data: None,
                    error: Some("missing_service".to_string()),
                }
            }
        };

        let text = message
            .get("content")
            .and_then(|c| c.get("text"))
            .and_then(|t| t.as_str())
            .unwrap_or("");

        // Try structured input from options
        if let Some(options) = message.get("options") {
            if let (Some(name), Some(schedule_str)) = (
                options.get("name").and_then(|n| n.as_str()),
                options.get("schedule").and_then(|s| s.as_str()),
            ) {
                let schedule = match parse_schedule(schedule_str) {
                    Ok(s) => s,
                    Err(e) => {
                        return ActionResult {
                            success: false,
                            text: format!("Invalid schedule: {}", e),
                            data: None,
                            error: Some(e),
                        }
                    }
                };

                let prompt = options
                    .get("prompt")
                    .and_then(|p| p.as_str())
                    .unwrap_or("Run scheduled task");

                let payload = PayloadType::Prompt {
                    text: prompt.to_string(),
                };

                return match service.create_job(
                    name.to_string(),
                    options.get("description").and_then(|d| d.as_str()).map(String::from),
                    schedule,
                    payload,
                    options.get("max_runs").and_then(|m| m.as_u64()),
                    options.get("room_id").and_then(|r| r.as_str()).map(String::from),
                ) {
                    Ok(job) => {
                        let schedule_str = format_schedule(&job.schedule);
                        ActionResult {
                            success: true,
                            text: format!(
                                "Created cron job \"{}\"\n- ID: {}\n- Schedule: {}\n- Status: {:?}",
                                job.name, job.id, schedule_str, job.state
                            ),
                            data: serde_json::to_value(&job).ok(),
                            error: None,
                        }
                    }
                    Err(e) => ActionResult {
                        success: false,
                        text: format!("Failed to create job: {}", e),
                        data: None,
                        error: Some(e),
                    },
                };
            }
        }

        // Natural language parsing
        let (name, schedule, prompt) = parse_create_request(text);

        let schedule = match schedule {
            Some(s) => s,
            None => {
                return ActionResult {
                    success: false,
                    text: "Could not understand the schedule. Try:\n- \"every 5 minutes\"\n- \"daily at 9am\"\n- A cron expression like \"0 9 * * 1-5\"".to_string(),
                    data: None,
                    error: Some("Could not parse schedule".to_string()),
                }
            }
        };

        let payload = PayloadType::Prompt {
            text: prompt.unwrap_or_else(|| "Run scheduled task".to_string()),
        };

        match service.create_job(name, None, schedule, payload, None, None) {
            Ok(job) => {
                let schedule_str = format_schedule(&job.schedule);
                ActionResult {
                    success: true,
                    text: format!(
                        "Created cron job \"{}\"\n- ID: {}\n- Schedule: {}\n- Status: {:?}",
                        job.name, job.id, schedule_str, job.state
                    ),
                    data: serde_json::to_value(&job).ok(),
                    error: None,
                }
            }
            Err(e) => ActionResult {
                success: false,
                text: format!("Failed to create job: {}", e),
                data: None,
                error: Some(e),
            },
        }
    }

    fn examples(&self) -> Vec<ActionExample> {
        vec![
            ActionExample {
                user_message: "Create a cron job to check the news every hour".to_string(),
                agent_response: "Created cron job \"check the news\"\n- ID: abc-123\n- Schedule: every 1 hour\n- Status: Active".to_string(),
            },
            ActionExample {
                user_message: "Schedule a daily reminder at 9am to review my goals".to_string(),
                agent_response: "Created cron job \"review my goals\"\n- ID: def-456\n- Schedule: cron: 0 9 * * *\n- Status: Active".to_string(),
            },
        ]
    }
}

/// Extracts name, schedule, and prompt from a natural language create request.
fn parse_create_request(text: &str) -> (String, Option<ScheduleType>, Option<String>) {
    let mut name = "Unnamed cron job".to_string();
    let mut prompt = None;

    // Extract "to <action>" pattern
    let to_re = regex::Regex::new(r"(?i)(?:to|that)\s+(.+?)(?:\s+every|\s+at\s+\d|$)").ok();
    if let Some(re) = to_re {
        if let Some(caps) = re.captures(text) {
            if let Some(m) = caps.get(1) {
                let action_text = m.as_str().trim();
                name = action_text.chars().take(50).collect();
                prompt = Some(action_text.to_string());
            }
        }
    }

    // Extract "called/named X"
    let name_re = regex::Regex::new(r#"(?i)(?:called|named)\s+["']?([^"']+)["']?"#).ok();
    if let Some(re) = name_re {
        if let Some(caps) = re.captures(text) {
            if let Some(m) = caps.get(1) {
                name = m.as_str().trim().to_string();
            }
        }
    }

    // Try to extract schedule from natural language
    let schedule = parse_natural_language_schedule(text);

    // If natural language failed, try raw schedule extraction
    let schedule = schedule.or_else(|| {
        let every_re =
            regex::Regex::new(r"(?i)every\s+(\d+\s*(?:seconds?|minutes?|hours?|days?|weeks?))")
                .ok()?;
        if let Some(caps) = every_re.captures(text) {
            return parse_natural_language_schedule(&format!("every {}", caps.get(1)?.as_str()));
        }
        None
    });

    (name, schedule, prompt)
}
